templates/extras/simp.md

```diff
@@ -1,111 +1,254 @@
 ```
 
-This lemma is then added to `simp`'s armoury. Note several things however.
+When the simplifier is attempting to simplify a term `T` it looks
+through the `simp` lemmas known to the system at that time, in a
+sensible way (there are over ten thousand of them currently in
+`mathlib`!), and if it runs into a lemma of the form `A = B` or `A ↔
+B`, for which `A` shows up in `T`, it rewrites the lemma at `T` to
+replace it with a `B`, and starts again.
 
-1) It might not be wise to make a random theorem into a simp lemma. Ideally the result has to be of
-a certain kind, the most important kinds being those of the form `A=B` and `A↔B`. Note however that
-if you want to add `fact` to `simp`'s weaponry, you can prove
+Note that the simplifier works in one direction only: if `A = B` is a
+`simp` lemma, then `simp` replaces `A`s with `B`s, but it doesn't
+replace `B`s with `A`s. Hence a `simp` lemma should have the property
+that its right hand side should be simpler than its left hand side. In
+particular, `=` and `↔` should not be viewed as symmetric operators in
+this situation. The following lemma
 
 ```lean
-@[simp] lemma my_lemma : fact ↔ true
+@[simp] lemma mul_right_inv_bad (a : G) : 1 = a * a⁻¹ := ...
+```
+
+would be a terrible `simp` lemma; the simplifier would then attempt to
+replace `1` with `a * a⁻¹`, which is usually not a sensible direction
+of travel. Furthermore, if both `mul_right_inv` and
+`mul_right_inv_bad` are tagged with `simp` then this would cause the
+simplifier to loop.
+
+However, `mathlib` has a linter, which one can call with `#lint` in a
+Lean file, and which will attempt to flag any problems with `simp`
+lemmas in the file.
+
+When making a new definition, it is very common to have `@[simp]`
+lemmas which train the simplifier how to use the definition in a
+sensible manner. An example of this is in mathlib's
+[data.complex.basic](https://github.com/leanprover-community/mathlib/blob/master/src/data/complex/basic.lean),
+which has almost 100 `simp` lemmas in. Theorems such as `@[simp] lemma
+add_re (z w : ℂ) : (z + w).re = z.re + w.re := rfl` are proved in that
+file, even though they are true by definition, because they are
+important `simp` lemmas. The very slick proof in that file that the
+complex numbers are a commutative ring uses `simp` at a key moment to
+reduce goals such as `(z + w).re = (w + z).re` into `z.re + w.re =
+w.re + z.re`, which is a statement about real numbers which the `ring`
+tactic can solve.
+
+## Basic usage 
+
+`simp` -- tries to simplify the goal using all `simp` lemmas known to
+          Lean at that time.
+
+`simp [h1, h2]` -- uses all `simp` lemmas and also `h1` and `h2`
+                   (which can either be local hypotheses or other lemmas which
+		   are not tagged `simp` for some reason).
+
+`simp [← h]` -- uses all `simp` lemmas, and also `h : A = B` but in
+                 the form `B = A` (so `simp` rewrites `B`s to `A`s)
+
+`simp [-h]` -- stops `simp` from using lemma `h`.
```

Comment:
Maybe this should say `simp [-thm]`, to make it clear that what goes there is not a local hypothesis but a defined theorem.

Reply:
Thanks.